### PR TITLE
56 yodkhatha assignment ii   service core

### PR DESCRIPTION
**Pull request by YoBiKhu**
Please approve the service stubs first, before approving this one. I am afraid of merge conflicts otherwise

### CODE REVIEW

File: services/entity/carbon-intensity-provider/api/openapi.yaml [hunk 1]
```diff
@@ -37,33 +37,7 @@ paths:
                     type: number
```
Comment (Sepia2023) [inline services/entity/carbon-intensity-provider/api/openapi.yaml:37]: Remove `openapi.yaml` file from `/services/entity/carbon-intensity-provider/api/openapi.yaml`. It only belongs to `/services/carbon-intensity-provider/api/openapi.yaml`.

File: services/entity/user-management/api/openapi.yaml [hunk 2]
```diff
@@ -21,24 +21,23 @@ paths:
             schema:
```
Comment (Sepia2023) [inline services/entity/user-management/api/openapi.yaml:21]: Discard changes to `/services/entity/user-management/api/openapi.yaml`. This file is to be removed here: [63 yodkhatha assignment i service stubs user management #67](https://github.com/informatik-mannheim/cmg-ss2025/pull/67#pullrequestreview-2813211546)

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# Binaries for programs and plugins
```
Comment (Sepia2023) [inline .gitignore:1]: Discard `.gitignore`. File is to be added in [57 yodkhatha assignment i service stubs #69](https://github.com/informatik-mannheim/cmg-ss2025/pull/69/files)

File: services/carbon-intensity-provider/main.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	// Preload fixed zones manually as test(for Assignment II)
+	service.AddOrUpdateZone("DE", 140.5)
+	service.AddOrUpdateZone("FR", 135.2)
+	service.AddOrUpdateZone("US-NY-NYIS", 128.9)
+
+	// Start server
+	router := api.NewHandler(service)
+	log.Println("Carbon Intensity Provider running on :8080")
+	log.Fatal(http.ListenAndServe(":8080", router))
+}
```
Comment (Sepia2023) [inline services/carbon-intensity-provider/main.go:23]: The current main() function starts the HTTP server with http.ListenAndServe, but doesn't support graceful shutdown. This means in-flight requests could be terminated abruptly during server shutdown.
I suggest to support graceful shutdown for this service.

File: services/carbon-intensity-provider/adapters/handler-http/handler.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	rtr     mux.Router
+}
+
+var _ http.Handler = (*Handler)(nil)
+
+func NewHandler(service ports.Api) *Handler {
+
+	h := Handler{service: service, rtr: *mux.NewRouter()}
+	h.rtr.HandleFunc("/carbon-intensity-provider/{id}", h.handleGet).Methods("GET")
+	h.rtr.HandleFunc("/carbon-intensity-provider", h.handleSet).Methods("PUT")
```
Comment (Sepia2023) [inline services/carbon-intensity-provider/adapters/handler-http/handler.go:23]: As of now, the API only exposes two GET endpoints:
`GET /carbon-intensity/{zone}` and `GET /carbon-intensity/zones.`

While having a service method like AddOrUpdateZone is useful for local initialization or testing, implementing a `PUT` endpoint to update zones externally is currently unnecessary and out of scope.

I suggest creating 2 different handler methods for both endpoints:
```suggestion
	h.rtr.HandleFunc("/carbon-intensity-provider/{id}", h.handleGetZone).Methods("GET")
	h.rtr.HandleFunc("/carbon-intensity-provider", h.handleGetAllZones).Methods("PUT")
```

File: services/carbon-intensity-provider/adapters/handler-http/handler.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	err = h.service.Set(carbonIntensityProvider, r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusCreated)
+}
```
Comment (Sepia2023) [inline services/carbon-intensity-provider/adapters/handler-http/handler.go:44]: Remove handle method as suggested in the prior comment.

File: services/carbon-intensity-provider/adapters/handler-http/handler.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+func (h *Handler) handleGet(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	carbonIntensityProvider, err := h.service.Get(vars["id"], r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	json.NewEncoder(w).Encode(carbonIntensityProvider)
+}
```
Comment (Sepia2023) [inline services/carbon-intensity-provider/adapters/handler-http/handler.go:55]: Prior comment suggested splitting this method into 2 handlers with methodnames `handleGetZone` and `handleGetAllZones`

File: services/carbon-intensity-provider/core/service_test.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
+package core
```
Comment (Sepia2023) [inline services/carbon-intensity-provider/core/service_test.go:1]: In the default folder `entity` I saw that the repository and notifier is mocked. I can't really tell if tests including those mocks should be implemented or not.

Consider adding another test method for `TestGetAvailableZones` with an empty list of zones.

File: services/carbon-intensity-provider/core/service.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"errors"
+
+	"github.com/informatik-mannheim/cmg-ss2025/services/carbon-intensity-provider/model"
+)
+
+// CarbonIntensityService is the concrete implementation of the CarbonIntensityProvider interface.
+type CarbonIntensityService struct {
+	zoneMap map[string]model.CarbonIntensityData
+}
```
Comment (Sepia2023) [inline services/carbon-intensity-provider/core/service.go:12]: You use a zoneMap struct to store all your zones. To me it looks like the whole repository layer is obsolete for you.
Your `repo.go` file defined 2 repo methods `Store` and `FindById`. Both methods are implemented but never called.
Maybe consider using the repo instead of a zoneMap struct or completely remove Repo's from your service?

**Review by Sepia2023 — CHANGES_REQUESTED** [review]
I added some comments based on my thoughts. Feel free to dm me if you have questions.

File: services/entity/carbon-intensity-provider/api/openapi.yaml (hunk 1 above)
Comment (YoBiKhu) [reply services/entity/carbon-intensity-provider/api/openapi.yaml:37]: Removed the folder and file

File: services/entity/user-management/api/openapi.yaml (hunk 2 above)
Comment (YoBiKhu) [reply services/entity/user-management/api/openapi.yaml:21]: Removed it also here now

File: .gitignore (hunk 3 above)
Comment (YoBiKhu) [reply .gitignore]: should not matter, but removed it anyways

File: services/carbon-intensity-provider/main.go (hunk 4 above)
Comment (YoBiKhu) [reply services/carbon-intensity-provider/main.go]: graceful shutdown implemented @Sepia2023

Comment (Sepia2023) [reply services/carbon-intensity-provider/main.go]: Looking good to me. Thank you for implementing it!

File: services/carbon-intensity-provider/adapters/handler-http/handler.go (hunk 5 above)
Comment (YoBiKhu) [reply services/carbon-intensity-provider/adapters/handler-http/handler.go]: added the suggestion

File: services/carbon-intensity-provider/core/service_test.go (hunk 8 above)
Comment (YoBiKhu) [reply services/carbon-intensity-provider/core/service_test.go:1]: You were right, I removed them and added the other test as well

File: services/carbon-intensity-provider/core/service.go (hunk 9 above)
Comment (YoBiKhu) [reply services/carbon-intensity-provider/core/service.go]: I use the repo.go file now for cleaner structure

**Comment by Sepia2023** [comment]
Following images show that the service does run locally on port 8080 and fulfills the requests as shown below.


image
image

**Review by Sepia2023 — APPROVED** [review]
